api/vignettes/pecanapi.Rmd

```diff
@@ -0,0 +1,233 @@
+You can do this by creating a `docker-compose.override.yml` file with the following contents in the root directory of the PEcAn source code:
+
+```yml
+version: "3"
+services:
+  postgres:
+    ports:
+      - 5432:5432
+```
+
+Here, the first port is the one used to access the database (can be any open port; most PostgreSQL applications assume 5432 by default), and the second is the port the database is actually running on (which will always be 5432).
+After making this change, reload the `postgres` container by running `docker-compose up -d`.
+To check that this works, open an R session and try to create a database connection object to the PEcAn database.
+
+```{r}
+con <- DBI::dbConnect(
+  RPostgres::Postgres(),
+  user = "bety",
+  password = "bety",
+  host = "localhost",
+  port = 5432
+)
+DBI::dbListTables(con)[1:5]
+```
+
+This code should print out five table names from the PEcAn database.
+If it throws an error, you have a problem with your database connection.
+
+The rest of this tutorial assumes that you are using this same database connection object (`con`).
+
+# Building a settings object
+
+The PEcAn settings object is a nested list providing parameters for the various actions performed by the PEcAn workflow, including the trait meta-analysis, processing input files, and running models.
+It can be created manually with a bunch of `list` calls.
+However, this is tedious and error-prone, so `pecanapi` provides several utilities that facilitate this process.
+
+We start with a blank list.
+
+```{r}
+settings <- list()
+```
+
+Let's start by adding a basic database configuration to this settings list.
+
+```{r}
+settings <- add_database(settings)
+settings
+```
+
+All `add_*` functions work by incrementally adding to an input settings object and returning a new modified settings object.
+The first argument of these functions is always the settings list, which gives these functions a consistent syntax and makes it easy to string multiple settings modifications together using the `magrittr` pipe (`%>%`), similar to `tidyverse` tabular data manipulations.
+
+The `add_database` function adds a sensible default configuration for the PEcAn database in the right place with the right names in the settings file.
+These defaults can, of course, be modified (see `?add_database`).
+
+Similarly, we can also add PFTs to the settings object with the `add_pft` function.
+
+```{r}
+settings <- add_pft(settings, "temperate.deciduous")
+settings
+```
+
+This adds the `temperate.deciduous` PFT to the appropriate spot in the settings hierarchy.
+Whereas `add_pft` adds a single PFT to the settings, `add_pft_list` can add a vector of PFTs.
+
+```{r}
+settings <- add_pft_list(settings, c("temperate.coniferous", "miscanthus"))
+settings
+```
+
+Notice that `add_pft` and `add_pft_list` both automatically augment the PFT number (`pft$constants$num`) as well.
```

Comment:
I modeled this after what the web interface currently does. Whenever I've done ED runs recently, regardless of the PFTs I select, the numbers were added sequentially. Here is a piece of a recent XML from a sucessful web-interface run using ED:

```xml
  <pfts>
    <pft>
      <name>Optics.Temperate_Early_Hardwood</name> 
      <constants>
        <num>1</num>
      </constants>
    </pft>
    <pft>
      <name>Optics.Temperate_Late_Conifer</name> 
      <constants>
        <num>2</num>
      </constants>
    </pft>
    <pft>
      <name>Optics.Temperate_Late_Hardwood</name> 
      <constants>
        <num>3</num>
      </constants>
    </pft>
    <pft>
      <name>Optics.Temperate_Mid_Hardwood</name> 
      <constants>
        <num>4</num>
      </constants>
    </pft>
  </pfts>
```

I know for a _fact_ that those are not the correct PFT numbers for those PFTs, yet everything worked anyway. That said, I agree that this is not how I think the code _should_ work. If we are pulling all parameters from the `history.csv` file and ignoring PFT numbers altogether, then we shouldn't be setting the `num` tag because that's misleading. Alternatively, if we _are_ relying on ED's internal PFT numbers to set defaults, then we should actually do that, and we don't need to also be pulling those defaults from the `history.csv` file.

Regardless, if only some models are using the `num` tag, I'm perfectly happy to remove the counter from the code and force the user to set them manually (which it sounds like they should be doing).

Reply:
correct, this is something that has been on my list of "things to fix" for a long time. One option is to let the user provide the pftnumber as an argument and do a quick check to see if it is already in use.

Reply:
Do we even need to add the constants$num for the workflow to run? AFAIK things should run fine without it.

Reply:
Per the above comment, I removed the counter altogether in 76c563714.